sensor/kubernetes/listener/resources/selector.go

```diff
@@ -1,45 +1,153 @@
 package resources
 
 import (
+	"math"
+
 	"k8s.io/apimachinery/pkg/labels"
 )
 
-// selector is a restricted version of labels.Selector
+// labelWithLen is label.Labels with added Len() function
+type labelsWithLen interface {
+	Has(label string) (exists bool)
+	Get(label string) (value string)
+	Len() uint
+}
+
+// selector is a restricted version of selectorWrap
 type selector interface {
-	Matches(labels.Labels) bool
+	Matches(labelsWithLen) bool
+	getSelector() labels.Selector
+	getNumLabels() uint
+	getMatchNil() bool
+}
+
+type labelWithLenImpl struct {
+	labels map[string]string
+}
+
+func (l labelWithLenImpl) Has(label string) bool {
+	_, exists := l.labels[label]
+	return exists
+}
+
+func (l labelWithLenImpl) Get(label string) string {
+	return l.labels[label]
+}
+
+func (l labelWithLenImpl) Len() uint {
+	return uint(len(l.labels))
+}
+
+func createLabelsWithLen(labels map[string]string) labelWithLenImpl {
+	return labelWithLenImpl{labels}
+}
+
+// SelectorWrap holds a selector and information allowing for additional checks before matching
+type selectorWrap struct {
+	selector  labels.Selector
+	numLabels uint
+	matchNil  bool
+}
+
+func (s selectorWrap) Matches(labels labelsWithLen) bool {
+	if s.numLabels > labels.Len() {
+		return false
+	}
+	if s.numLabels == 0 {
+		return s.matchNil
+	}
+	return s.selector.Matches(labels)
+}
+
+func (s selectorWrap) getSelector() labels.Selector {
+	return s.selector
+}
+
+func (s selectorWrap) getNumLabels() uint {
+	return s.numLabels
+}
+
+func (s selectorWrap) getMatchNil() bool {
+	return s.matchNil
 }
 
 // selectorDisjunction is the disjunction (logical or) of a list of selectors.
-type selectorDisjunction []selector
+type selectorDisjunction []labels.Selector
 
-func (d selectorDisjunction) Matches(labels2 labels.Labels) bool {
+func (d selectorDisjunction) Empty() bool {
 	for _, sel := range d {
-		if sel.Matches(labels2) {
+		if !sel.Empty() {
+			return false
+		}
+	}
+	return true
+}
+
+func (d selectorDisjunction) String() string {
+	panic("unused function")
+}
+
+func (d selectorDisjunction) Add(r ...labels.Requirement) labels.Selector {
+	panic("unused function")
+}
+
+func (d selectorDisjunction) Requirements() (requirements labels.Requirements, selectable bool) {
+	panic("unused function")
+}
+
+func (d selectorDisjunction) DeepCopySelector() labels.Selector {
+	panic("unused function")
+}
+
+func (d selectorDisjunction) RequiresExactMatch(label string) (value string, found bool) {
+	panic("unused function")
+}
```

Comment:
Why do we need to implement `labels.Selector` (I am guessing that this is the reason why you added those functions)?

I think we should either remove them or implement them.

Reply:
The reason we need to implement labels.Selector is the way I handle disjunctions. A selectorWrapper holds a labels.Selector internally, or, in this case it holds a selectorDisjunction since it implements labels.Selector.

I could instead make a selectorDisjunction that implements selector and simply holds []selectorWrapper then forwards Matches() calls to all of them. That way the selectorDisjunction wouldn't have to implement labels.Selector and the code would become a bit shorter. Is that preferable?